Backup and restore stubbed method to and from prepended module

texpert opened the pull request:
Define the stub method directly on the last prepended method on which the target method has been defined..

These changes helped to fix the test suite of a project I'm working on, where I've started using prepended modules. Ruby 2.3.7 and Rails 4.1.16

godfat committed:
Fix stubbing prepended only methods

Previously, we're assuming the method must be defined in the
singleton class. However this is not always true. Whenever the
method was only defined in the prepended module, then it's not
defined in the singleton class. We need to find the owner of
the method instead, which is the prepended module.

Closes https://github.com/rspec/rspec-mocks/issues/1213

godfat committed:
Fix the case for extend self + any instance of

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -217,7 +220,7 @@ def restore_original_method!(method_name)
             #
             # https://github.com/ruby/ruby/commit/c8854d2ca4be9ee6946e6d17b0e17d9ef130ee81
-            if RUBY_VERSION < "2.3" || backed_up_method_owner[method_name.to_sym] == self
+            if RUBY_VERSION < "2.3" || method_owner == self
```
I do not really understand what this was trying to do, but the patch you're providing is simply making this always equals to true, from what I can see, because `self` is in the context of `method_owner`, therefore `method_owner` of course would be `self`.

The original code was comparing `method_owner` with `@klass`, which is a different case.

texpert replied:
Yeah, I think this check could be removed at all.

godfat replied:
Well, I think that's because you changed the aliased method to be on the method owner, therefore it must be coming from the method owner. However, originally it's from the `@klass` which might or might not be the method owner.

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -234,9 +237,11 @@ def backup_method!(method_name)
-          @klass.class_exec do
+          method_owner = @klass.instance_method(method_name).owner
+          @backed_up_method_owner[method_name.to_sym] ||= method_owner
+          method_owner.class_exec do
```
I think this can be wrong too. The method owner might be from a module, but we won't want to touch the module, but the `@klass`

texpert replied:
I have debugged this thoroughly. Prior to this PR:

- the method was aliased on the @klass, but not removing the original method, so stubbing wasn't working - the original method stayed in place
- trying to add `remove_method method_name` after the `alias_method alias_method_name, method_name` brought the `method `sync_product_to_flow' not defined in #<Spree::Variant not initialized>`

That's why I have decided that it is better to alias and remove the original method on its owner. The stub is created directly on the @klass, and removed from the @klass too, while restoring the original method on the owner.

godfat replied:
I'll take a closer look tomorrow, but want to let you know the case I have in mind. What would happen here?

``` ruby
module M
  def f
    'f'
  end
end

class C0
  include M
end

class C1
  include M
end

c0 = C0.new
c1 = C1.new

allow(c0).to receive(:f).and_return('g')

c0.f # => g
c1.f # => f?
```

texpert replied:
Yeah, the method will be changed for both classes, for sure. But then, to make it work for just one class, there should be some more work done on stubbing - like the new method analyzing the context and switching to original method if it finds itself in the non-target class.

godfat replied:
I just tested. The current behaviour will give `f` for `c1.f`, so I don't think this is acceptable as it's changing the behaviour, and I don't think it should change for both classes :(

If we're stubbing on `C0` not on method owner, it won't affect `C1`. So I think this is the wrong fix. I suggest you put some simple tests to show what we're trying to fix here. Otherwise I don't know what I can suggest beside pointing out that this can change the behaviour which is not desirable.

texpert replied:
Oh, my bad, I thought of `allow_any_instance_of` instead of `allow`.

It is working OK with instance stubbing.
<img width="1680" alt="Screenshot 2020-12-04 at 20 12 17" src="https://user-images.githubusercontent.com/2990033/101199464-95f30900-366d-11eb-8f3a-7ab3e982a626.png">

texpert replied:
But with `allow_any_instance_of `, we got `c1.f undefined method `f' for #<C1:0x00007fcd8b06f350>`

godfat replied:
> the method was aliased on the @Klass, but not removing the original method, so stubbing wasn't working - the original method stayed in place

I am confused. Why should it remove the original method? Can you show a specific example?

>  But with allow_any_instance_of , we got c1.f undefined method f' for #C1:0x00007fcd8b06f350`

I tried `allow_any_instance_of` on both `C1`, `C2`, and `M`. Without your patch it's all working as expected to me. So I am still not clear what you're trying to fix, and I don't see the error you're mentioning without your patch.

godfat reviewed (commented):
I have some troubles running the test suite so I didn't try to verify this yet, but would you be able to put a test here which can prove that this will fix something? Beside that I have some questions for you. I commented inline.

And just to be clear, this is not the official rspec-mocks repository, so even if I merge it it's unlikely to go into the official branch soon.

texpert commented:
> I have some troubles running the test suite so I didn't try to verify this yet, but would you be able to put a test here which can prove that this will fix something? 

I have not managed to set properly the test suite. Have tried to change the gemspec and Gemfile locally with some hardcoded versions for rspec gems, but got a suite, that was giving spurious test failures - each time different - from 8 to a few hundreds of failures :(

> And just to be clear, this is not the official rspec-mocks repository, so even if I merge it it's unlikely to go into the official branch soon.

That's not a problem for my project - I have forked you repo with this branch, implemented my changes, and our test suite is doing well.

godfat reviewed (commented):
I'll try to rebase at some point, and it might fix the test suite. However I would like to know what you're trying to fix first.

texpert commented:
> I'll try to rebase at some point, and it might fix the test suite. However I would like to know what you're trying to fix first.

I am trying to fix working of `allow_any_instance_of` when prepending the modules (when including the modules it works OK, yes):

```
require 'rails_helper'

class C0
end

module M
  def f
    'f'
  end

  ::C0.prepend(self) if ::C0.included_modules.exclude?(self)
end

class C1
  include M
end

c0 = C0.new
c1 = C1.new

RSpec.describe C0 do
  it do
    # allow(c0).to receive(:f).and_return('g')
    allow_any_instance_of(C0).to receive(:f).and_return('g')

    c0.f # => g
    c1.f # => f?
  end
end
```

Without the patch is failing:

```
0) C0 
     Failure/Error: allow_any_instance_of(C0).to receive(:f).and_return('g')
       Using `any_instance` to stub a method (f) that has been defined on a prepended module (M) is not supported.
     # ./spec/services/prepend_test_spec.rb:24:in `block (2 levels) in <top (required)>'
```

With the patch the `allow_any_instance_of` is working OK, but we got c1.f undefined method f' for #C1:0x00007fcd8b06f350`.

godfat commented:
Thank you for the simple test which I can verify. I understand what you're trying to fix now, but I believe it's the wrong fix. I tried the same test on my mocking library, [Muack](https://github.com/godfat/muack), and it's not working as expected either.

Both `c0.f` and `c1.f` would give `'g'`. 

I think I understand why it's happening. The problem here is that the prepended module comes first than the singleton class, therefore it's acting on the prepended module `M` here, rather than the singleton class of `C0`.

In order to fix this, we'll need to add something before the prepended module, which, could be another generated prepended module. This is sad because there's no way to remove the prepended module therefore it means the class cannot be reset back to the original state.

But maybe we can minimize the damage by making it a truly transparent module? I have no answer yet. I'll see if I can fix this in my mocking library first, and see if that knowledge can be applied back to rspec-mocks because I am much more familiar with my mocking library, making it much easier to experiment potential solution.

godfat commented:
So I committed https://github.com/godfat/muack/commit/dff0842139b06ac0dcc72599a05f84b421791a9b which should fix most of the use cases. It is essentially doing exactly what I described above: Prepend an additional module whenever we detected that there's a module being prepended and on the way.

``` ruby
def self.prepare_target singleton_class
  target = singleton_class.ancestors.first
  if target == singleton_class # no prepended module detected
    target
  else # prepend an internal module to override the prepended module(s)
    name = :MuackPrepended
    if singleton_class.const_defined?(name)
      singleton_class.const_get(name)
    else
      prepended = ::Module.new
      singleton_class.const_set(name, prepended)
      singleton_class.private_constant(name)
      singleton_class.prepend(prepended)
      prepended
    end
  end
end
```

And we mock on the newly prepended module to "override" the prepended module. Maybe this can be more precise that this is only needed when the method name was conflicted. I'll see if I can improve this part and see how that would be different.

I think the same idea can be applied to rspec-mocks but I am not fully confident where it should be applied and if this can be accepted. I will try to get there at some point, but no promise though, since I don't really want to maintain rspec-mocks.

texpert commented:
Great!

I have been exactly thinking of dynamically prepending a Module, as per your suggestion 7 hours ago.

Maybe even not need to analyze is the method from a prepended Module, or not, just always prepend the stub?

texpert commented:
@godfat , could you, please, rebase the https://github.com/rspec/rspec-mocks/pull/1218 on a fresh `main`so maybe to fix the test suite?

godfat commented:
> Maybe even not need to analyze is the method from a prepended Module, or not, just always prepend the stub?

@texpert I think it would be much better to only prepend a module when it's absolutely needed, because there's no way we can remove this newly prepended module and that's changing the application code for the lifetime of the process, which can be breaking application codes which might be relying on ancestors chain being intact.

For example, this tool will be quite sensitive to that: https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/utils/override.rb

I just pushed another commit to make this conditional: https://github.com/godfat/muack/commit/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc The critical check is:

``` ruby
index = singleton_class.ancestors.index(singleton_class)
prepended_modules = singleton_class.ancestors[0...index]

if prepended_modules.find{ |m| direct_method_defined?(m, msg) }
  # .. an internal prepend is needed
else
  # can just use the singleton class
end
```

There's only one case which I think it's probably impossible to solve right now:

``` ruby
m0 = Module.new{ def f; :m0; end }
m1 = Module.new{ def f; :m1; end; prepend m0 }
c0 = Class.new{ prepend m0 }.new
c1 = Class.new{ prepend m1 }.new

mock(any_instance_of(m1)).f{:g}

expect(c0.f).eq :m0
expect(c1.f).eq :g # m1 does not know c1 thus no way to pass this
```

I am skipping this test for now, but I hope in the future it's possible to solve this.

> could you, please, rebase the rspec#1218 on a fresh mainso maybe to fix the test suite?

Yes, I have rebased it, and it's great to see that all tests just passed now, and I didn't even change anything.

zhisme committed:
raise ArgumentError if object cannot be proxied

zhisme committed:
more detailed error message, improve tests

texpert committed:
Backup and restore stubbed method to and from prepended module

texpert committed:
Fix rubocop issue

texpert commented:
I've managed to implement stubbing the last prepended module, so no need to create dynamically another one prepended :)

JonRowe committed:
Updated ci build scripts (from rspec-dev)

godfat commented:
> I've managed to implement stubbing the last prepended module, so no need to create dynamically another one prepended :)

What about this case?

``` ruby
module M
  def f
    :f 
  end
end

class C0
  prepend M
end

class C1
  prepend M
end

RSpec.describe do
  it do
    allow_any_instance_of(C0).to receive(:f).and_return(:g)

    expect(C0.new.f).to eq(:g)
    expect(C1.new.f).to eq(:f)
  end
end
```

If you're stubbing the last prepended module, I would guess now stubbing on `C0` would affect `C1`, which is not desired.

That's where why we need to prepend a new module, because the prepended module can be used by the completely unrelated class.

JonRowe committed:
Merge pull request #1360 from rspec/update-ci-build-scripts-2020-12-08-for-main

Updates from rspec-dev (2020-12-08)

JonRowe committed:
Updated ci build scripts (from rspec-dev)

texpert committed:
Allow stubbing methods defined on the same prepended module for different classes

texpert committed:
Fix rubocop issue

texpert committed:
Call the original method for the not-stubbed classes

texpert commented:
Seems that I've managed to do it, finally, for all the cases - added few specs for these cases too.

JonRowe committed:
Merge pull request #1362 from rspec/update-ci-build-scripts-2020-12-09-for-main

Updates from rspec-dev (2020-12-09)

zhisme committed:
add support for ruby < 2.0 symbols not frozen

godfat commented:
Hopefully I'll review tomorrow.

zhisme committed:
unified solution for all ruby versions

JonRowe committed:
Merge pull request #1357 from zhisme/main

raise ArgumentError if object cannot be proxied

JonRowe committed:
Change log for #1357

JonRowe committed:
Updated ci build scripts (from rspec-dev)

JonRowe committed:
Merge pull request #1366 from rspec/update-ci-build-scripts-2020-12-21-for-main

Updates from rspec-dev (2020-12-21)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main-

JonRowe committed:
Merge pull request #1369 from rspec/update-ci-build-scripts-2020-12-22-for-main-

Updates from rspec-dev (2020-12-22)

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -123,6 +124,14 @@ def build_alias_method_name(method_name)
+        # @private
+        def build_alias_prepended_method_name(method_name)
+          unless @klass.name
+            Object.const_set("Klass_#{@klass.object_id}", @klass)
```
Do we really need this? Setting a constant at a global namespace feels wrong to me, and we should somehow clean this up in the cleanup process.

texpert replied:
Cleaning implemented in https://github.com/godfat/rspec-mocks/pull/1/commits/70c2b88f6295361ed33a5a33959babc32d919a01

godfat replied:
I'll prefer if this can be set inside a non-top level namespace, for example:

``` ruby
RSpec::Mocks.const_set("Klass_#{@klass.object_id}", @klass)
```

But this is less important.

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -123,6 +124,14 @@ def build_alias_method_name(method_name)
+          unless @klass.name
+            Object.const_set("Klass_#{@klass.object_id}", @klass)
+          end
+          "__#{@klass.name}_#{method_name}_without_any_instance__"
```
We can probably just use `"Klass_#{@klass.object_id}"` as the class name here because they're the same anyway, and we avoid touching the global namespace.

``` ruby
"__Klass_#{@klass.object_id}_#{method_name}_without_any_instance__"
```

texpert replied:
Nope, they're not the same. This is in the case of dynamically defined classes without a name - the object id's are changing, the class name remains the same. That's why I had to name somehow the class, if it has no name.

godfat replied:
I should have responded to this earlier. Sorry for that.

I am confused why they are not the same, specifically given this line:

> the object id's are changing, the class name remains the same

Why is the id changing, yet its name remains the same? Here if `@klass.name` does not exist, meaning that it's a dynamically generated module, which we set the name based on `@klass.object_id`, and then we use `@klass.name`, doesn't that mean:

``` ruby
"__#{@klass.name}_#{method_name}_without_any_instance__"
```

is equivalent to:

``` ruby
"__Klass_#{@klass.object_id}_#{method_name}_without_any_instance__"
```

?

Could you please give an example when id would be changing, yet the name is the same? Do you mean when the user explicitly sets a name? In that case, I am not sure what we should really do. My thinking is, if the id is changing, doesn't mean they are different things? So they shouldn't share the same thing?

godfat commented on spec/rspec/mocks/any_instance_spec.rb:
```diff
@@ -176,12 +180,44 @@ def private_method; :private_method_return_value; end
+          end
+
+          context 'different classes having the same module prepended' do
+            let(:klass1) do
```
Should probably be called something else, because `klass1` is a bit confusing. Maybe `other_klass`.

texpert replied:
Fixed in https://github.com/godfat/rspec-mocks/pull/1/commits/e30bf39612257c00b2f678f2d0f21a3cf4b032e4

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
-            # nothing to do; prepends aren't supported on this version of ruby
-          end
+        def prepended_module(method_name)
+          @prepended_module =
```
I can't tell if this might be causing a problem for now, but I think this is too stateful and could easily lead to bugs. Consider this case:

``` ruby
prepended_module(:test0)
@prepended_module # this is based on :test0
prepended_module(:test1)
@prepended_module # this is based on :test1
```

This memorization should be based on `method_name`, or we just don't memorize at all.

texpert replied:
Hm, but it IS based on method_name.

Could you, please, elaborate waht do you mean or how to test it?

godfat replied:
> Hm, but it IS based on method_name.

When you're accessing `@prepended_module`, it's not based on `method_name`, unless you're always using `prepended_module(method_name)` so you're always passing `method_name`, but then this memorization is always discarded and has no effect.

> Could you, please, elaborate waht do you mean or how to test it?

In practice, it can be possible that this disparity might not happen, which I don't know if it'll happen or not, but I think this is a code smell because the assumption is not consistent. When you access `@prepended_module`, you'll need to know which was the last time `prepended_module(method_name)` was called and if that `method_name` is matching the one you have in mind for this time. It's very easy to get wrong with this dependency.

texpert replied:
OK, @godfat , I've added a test for stubbing methods defined in different prepended modules, and, also, added fetching the exact prepended_module where the method is defined in `restore_original_method!` method.

Is it good now?

godfat replied:
@texpert Sorry for not responding for so long. There were some bigger holidays and I was getting a bit more busy lately so I didn't look into this for quite some time. I'll get onto this again soon (in a few days maybe?), and definitely when the pull request you're basing off got merged upstream. It's sad that it takes so long and it's still not there yet, but I guess that's it. (as part of my excuse :s)

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
-          end
+        def prepended_module(method_name)
+          @prepended_module =
+            RSpec::Mocks::Proxy.prepended_modules_of(@klass).select { |mod| mod.method_defined?(method_name) }.last
```
Three things here.

0. `method_defined?` does not cover private methods, so `private_method_defined?` should also be used.
1. It should not look for super methods, therefore `false` should be passed to `method_defined?` too. Unfortunately, older Rubies do not accept more arguments, we'll need to fall back to `instance_methods(false).include?(method_name)`
2. Why is it on `last` prepended module? What would happen if there are multiple prepended modules which all have the same method?

For the second point, consider this case:

```
>> Class.new.method_defined? :to_s
=> true
>> Class.new.method_defined? :to_s, false
=> false
```

How I implemented it in Muack: https://github.com/godfat/muack/blob/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc/lib/muack/mock.rb#L151-L154

``` ruby
if ::Class.instance_method(:method_defined?).arity == 1 # Ruby 2.5-
  def self.direct_method_defined? mod, msg
    mod.public_instance_methods(false).include?(msg) ||
      mod.protected_instance_methods(false).include?(msg) ||
      mod.private_instance_methods(false).include?(msg)
  end
else
  def self.direct_method_defined? mod, msg
    mod.method_defined?(msg, false) || # this doesn't cover private method
      mod.private_method_defined?(msg, false)
  end
end
```

For the third point, consider this case:

``` ruby
mod0 = Module.new{ def f; :m0; end }
mod1 = Module.new{ def f; :m1; end }
klass0 = Class.new{ prepend mod0 }
klass1 = Class.new{ prepend mod1 }
klass = Class.new{ prepend mod1; prepend mod0 }

allow_any_instance_of(klass).to receive(:f).and_return(:f)

expect(klass.new.f).to eq(:f)
expect(klass0.new.f).to eq(:m0)
expect(klass1.new.f).to eq(:m1)
```

texpert replied:
Oops, as for why the `last` method - it was just a dumb semantic error - sure I've meant `first`, as we're dealing with a prepended module.

Fix in https://github.com/godfat/rspec-mocks/pull/1/commits/a352272418ce91d21e2ef1a8876001dfdd185168 and your case added to the specs.

Thanks a lot!!

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -257,9 +288,31 @@ def observe!(method_name)
+              @prepended_module.__send__(:define_method, method_name) do |*args, &blk|
+                possible_method_stub = "__#{self.class.name}_#{method_name}_without_any_instance__"
+                if respond_to?(possible_method_stub)
+                  __send__(possible_method_stub, *args, &blk)
```
Interesting approach! So far I can't really think of a downside for this, but I'll keep pondering if I can find a case where this cannot cover.

godfat replied:
Here's a test case for you:

``` ruby
mod = Module.new{ def f; :f; end }
base = Class.new{ prepend mod }
sub = Class.new(base)

allow_any_instance_of(base).to receive(:f).and_return(:g)

expect(base.new.f).to eq(:g)
expect(sub.new.f).to eq(:g)
```

This may show that `self.class` is not a sufficient key. The same goes to stubbing on `mod`, which all other instances might have all different classes yet since we're stubbing on `mod` it should affect all of them.

texpert replied:
I have fixed this too - https://github.com/godfat/rspec-mocks/pull/1/commits/db750748f0f5ebddbf9d53b83b907d741bf60268

godfat replied:
You covered my point 2, but didn't cover my point 0 and 1.

godfat reviewed (commented):
Interesting approach. I left some comments about some implementation details, which would be nice to have some covering tests but I guess that is ok. I'll keep thinking on some tricky edge cases and see if this will work for all of them.

godfat commented:
And sorry for the delay! I went on vacation and didn't really want to review this while I was taking the vacation 😅

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -275,18 +328,9 @@ def mark_invoked!(method_name)
-          def allow_no_prepended_module_definition_of(_method_name)
-            # nothing to do; prepends aren't supported on this version of ruby
-          end
+        def prepended_module(method_name)
```
My solution for this on Muack is this: https://github.com/godfat/muack/blob/e8bd6473ae58b805cbc0101fb2bc14eff9c379fc/lib/muack/mock.rb#L167-L190

``` ruby
def self.prepare_target singleton_class, msg
  if singleton_class == singleton_class.ancestors.first # no prepended mod
    singleton_class
  else # check if we need to prepend an internal module to override
    index = singleton_class.ancestors.index(singleton_class)
    prepended_modules = singleton_class.ancestors[0...index]

    if prepended_modules.find{ |m| direct_method_defined?(m, msg) }
      # prepend an internal module to override the prepended module(s)
      name = :MuackPrepended
      if singleton_class.const_defined?(name)
        singleton_class.const_get(name)
      else
        prepended = ::Module.new
        singleton_class.const_set(name, prepended)
        singleton_class.private_constant(name)
        singleton_class.prepend(prepended)
        prepended
      end
    else # we don't need to override so singleton class is enough
      singleton_class
    end
  end
end
```

I think this is not perfect because we'll be inserting an internal prepended module, but so far this solved any edge cases I can think of. I hope this can give you some inspiration.

texpert commented:
Thank you for review and hints, @godfat ! Will try to further work on the PR as soon, as I'll get some time.

benoittgt committed:
Updated common plaintext files (from rspec-dev)main-

JonRowe committed:
Updated common plaintext files (from rspec-dev)main

JonRowe committed:
Merge pull request #1374 from rspec/update-common-plaintext-files-2020-12-25-for-main

Updates from rspec-dev (2020-12-25)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1375 from rspec/update-ci-build-scripts-2020-12-26-for-main

Updates from rspec-dev (2020-12-26)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1378 from rspec/update-ci-build-scripts-2020-12-27-for-main

Updates from rspec-dev (2020-12-27)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1381 from rspec/update-ci-build-scripts-2020-12-27-for-main

Updates from rspec-dev (2020-12-27)

JonRowe committed:
v3.10.1

texpert committed:
Clean @klass.name constant, if it was created on stubbing

texpert committed:
Change name klass1 to other_klass for clarity.

texpert committed:
Fix to stub the first of ancestors, not last, when stubbing a method on a prepended module

texpert committed:
Fix rubocop styling issue

texpert committed:
Fix to find and use the correct stub in case a method is defined in a module prepended to a superclass, searching for the method on all the ancestor chain

texpert commented:
I have still to work on restoring the original method visibility, but this will be another day when I'll got some time.

Commit:
Mark proxy methods with ruby2_keywords

Commit:
Only use ruby2_keywords if defined

Commit:
Only test keyword arguments on rubies that support them

Commit:
Update a spec to work in rubies that don't support required kwargs

pirj committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1388 from rspec/update-ci-build-scripts-2020-12-31-for-main

Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1391 from rspec/update-ci-build-scripts-2021-01-02-for-main

Updates from rspec-dev (2021-01-02)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Merge pull request #1392 from rspec/update-ci-build-scripts-2021-01-02-for-main

Updates from rspec-dev (2021-01-02)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -275,18 +336,9 @@ def mark_invoked!(method_name)
-          end
+        def prepended_module(method_name)
+          @prepended_module =
+            RSpec::Mocks::Proxy.prepended_modules_of(@klass).select { |mod| mod.method_defined?(method_name) }.first
```
I should probably mention this earlier, but only now I realized. For `select` and then `first`, that's `find`.

So this can be just:

``` ruby
RSpec::Mocks::Proxy.prepended_modules_of(@klass).find { |mod| mod.method_defined?(method_name) }
```

texpert replied:
Fixed in https://github.com/godfat/rspec-mocks/pull/1/commits/31768dd8659c23c2263ea2135e4998f879fed9bf

JonRowe committed:
Merge pull request #1393 from rspec/update-ci-build-scripts-2021-01-02-for-main

Updates from rspec-dev (2021-01-02)

Commit:
Avoid respond_to? when checking existence of the ruby2_keywords method

Using `respond_to?` causes issues in rspec-rails because of how the `Railtie` class implements `respond_to_missing?`.

JonRowe committed:
Merge pull request #1385 from bryanp/fix-kwargs-for-ruby3

Resolve an issue using method doubles with keyword arguments on Ruby 3

JonRowe committed:
Change log for #1385

pirj committed:
Use true/false for previously_defined

```
  1) RSpec::Mocks::Constant.original for a previously defined unstubbed constant indicates it was previously defined
     Failure/Error: it("indicates it was previously defined") { expect(const).to be_previously_defined }
       expected `#<RSpec::Mocks::Constant TestClass::M>.previously_defined?` to return true, got [:m, "TestClass::M"]
     # ./spec/rspec/mocks/mutate_const_spec.rb:446:in `block (4 levels) in <module:Mocks>'
```

JonRowe committed:
Changelog for #1397

JonRowe committed:
Merge pull request #1402 from rspec/backport-strict-predicate-fix

Backport strict predicate fix

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Pin childprocess

JonRowe committed:
Add coverage group

JonRowe committed:
Pin thor

JonRowe committed:
Merge pull request #1403 from rspec/update-ci-build-scripts-2021-01-22-for-main

Updates from rspec-dev (2021-01-22)

JonRowe committed:
Updated ci build scripts (from rspec-dev)main

JonRowe committed:
Remove .travis.yml

JonRowe committed:
Merge pull request #1405 from rspec/update-ci-build-scripts-2021-01-22-for-main

Updates from rspec-dev (2021-01-22)

JonRowe committed:
Update build status

jhottenstein committed:
Add tests and fix for ruby 3 kwargs issues

jhottenstein committed:
Add tests (and feature sniffing) for mocking kwarg methods

jhottenstein committed:
Fix jruby test by removing the receiver from the block args

JonRowe committed:
Merge pull request #1407 from jhottenstein/jess/add-kwarg-support-for-any-instance

Add tests and fix for ruby 3 keyword arguments issues

JonRowe committed:
Changelog for #1407
[skip ci]

JonRowe committed:
v3.10.2

texpert committed:
Fix `select ... first` to `find` in RSpec::Mocks::AnyInstance::Recorder.prepended_module

texpert committed:
Mark defined methods taking undefined args as ruby2keywords when ruby2keywords is defined in RSpec::Mocks::AnyInstance::Recorder.prepended_module

texpert commented:
> I have still to work on restoring the original method visibility, but this will be another day when I'll got some time.

I have overlooked that everything's OK with preserving and restoring method's visibility.

texpert committed:
Correct some prepend test with properly defined modules

texpert committed:
Add test for stubbing a chain of methods defined on different prepended module

texpert committed:
Get the exact prepended_module where the method is defined in `restore_original_method!` method

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
+          if prepended_module(method_name)
+            alias_prepended_method_name = build_alias_prepended_method_name(method_name)
+            module_methods = @prepended_module.public_instance_methods
+            remaining_stubs = module_methods.select { |m| m.end_with?("_#{method_name}_without_any_instance__") }
```
A new idea here. First we can use `count` instead of `select` because it seems we don't really need those name, just how many there are.

Secondly, can we maybe count it internally, instead of observing it?

Every time we insert a new stub, we do `@remaining_stubs += 1` and every time we remove a stub, we do `@remaining_stubs -= 1`.

Generally, checking by counting names, which can easily hit into name conflicts, gives me some nerve.

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
+              remove_method alias_prepended_method_name
+              Object.send(:remove_const, @klass.name) if @klass.name.start_with?('Klass_')
+
+              break if remaining_stubs.size > 1
```
Could you please explain why we're always removing the module, but not always removing the method? (only when `remaining_stubs.size > 1`)

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -196,31 +205,48 @@ def restore_original_method!(method_name)
               alias_method method_name, alias_method_name
+              remove_method alias_method_name
+            end
+            @prepended_module = nil if remaining_stubs.size == 1 # This was the last stub
```
Shouldn't we remove the module only at this point?

godfat commented on lib/rspec/mocks/any_instance/recorder.rb:
```diff
@@ -235,8 +261,17 @@ def backup_method!(method_name)
+            unless @prepended_module.method_defined?(alias_method_name)
+              @prepended_module.module_exec do
+                alias_method alias_method_name, method_name
+                remove_method method_name
```
Why do we only remove method in this case (when there is `prepended_module`), but not for the other case?

godfat commented:
@texpert I am very sorry for the long delay, and I have less free time recently. Given that it's been awhile I am sure I must have forgotten some of the details and was probably missing a lot of context in my head. Apologies if some of my questions look a bit silly or out of places. 

To be honest, I am feeling a bit tired of reviewing patches against codes not maintained by me. I don't really understand how those work, thus it's difficult to review patches against them. In order to properly give a proper review, I have to understand the original codes first, and then I can continue to understand the patch.

I've done this in the past few rounds of reviews, which really took a lot of time. I think I'd done the right thing, but given that my pull request isn't even merged yet, I am not fully sure if the effort spent upfront is justified.

I think I will pause on very thorough review, and only give some quick read and review from now on. And I would really like to pass the review to the RSpec team, if possible.

I thank you for contributing patches based on my patch, I am honored by that. Let's hope https://github.com/rspec/rspec-mocks/pull/1218 can be merged soon and we can move on from there.

godfat commented:
@texpert Now the base branch is merged into `4-0-dev` upstream, do you want to rebase and create a new pull request at https://github.com/rspec/rspec-mocks and also target `4-0-dev`, so that RSpec team members can take a look?

texpert commented:
> @texpert Now the base branch is merged into `4-0-dev` upstream, do you want to rebase and create a new pull request at https://github.com/rspec/rspec-mocks and also target `4-0-dev`, so that RSpec team members can take a look?

Sure, @godfat - will find time either this, or next weekend.

Congratulations with the merge of your PR!

godfat commented:
Thanks! I'll try to review your PR as well. Hopefully we can fix all the
wild corner cases.
